real_preprocessing/src/corner_detection.cpp

```diff
@@ -0,0 +1,104 @@
+#include "ros/ros.h"
+#include "apriltag_ros/AprilTagDetectionArray.h"
+#include "tf_conversions/tf_eigen.h"
+#include "Eigen/Dense"
+#include "fstream"
+#include "string"
+#include "iostream"
+
+int count(0);
+std::string capture("1");//prevenets first iteration YAML dump
+std::ofstream fout;
+
+void aprilDetection(const apriltag_ros::AprilTagDetectionArray::ConstPtr& msg)
+{ 
+  if(capture=="") //specifies trigger key to dump YAML
+  {
+    if (!msg->detections.empty())
+    {
+      //intrinsic specification
+      Eigen::MatrixXd intrinsic(3, 3);  
+      intrinsic << 630.3467672874693, 0, 314.10832937485145, 0, 624.9044125522948, 241.59156511756711, 0, 0, 1;
+      
+      //file name generation
+      std::string filename("detections_"+std::to_string(count)+".yml");  
+      fout.open(filename.c_str());
+      fout << "detections:";
+
+      for (int i = (msg->detections.size() - 1); i > -1; i--)
+      {
+        // reading data from topic
+        float cam_T_tag_xpos = msg->detections[i].pose.pose.pose.position.x;
+        float cam_T_tag_ypos = msg->detections[i].pose.pose.pose.position.y;
+        float cam_T_tag_zpos = msg->detections[i].pose.pose.pose.position.z;
+        float cam_T_tag_xori = msg->detections[i].pose.pose.pose.orientation.x;
+        float cam_T_tag_yori = msg->detections[i].pose.pose.pose.orientation.y;
+        float cam_T_tag_zori = msg->detections[i].pose.pose.pose.orientation.z;
+        float cam_T_tag_wori = msg->detections[i].pose.pose.pose.orientation.w;
+        double tag_size = msg->detections[i].size[0];
+        int tag_id = msg->detections[i].id[0];
+
+        tf::Quaternion q(cam_T_tag_xori, cam_T_tag_yori, cam_T_tag_zori, cam_T_tag_wori);
+        tf::Matrix3x3 m(q);  // quaternion to rotational matrix
+
+        Eigen::MatrixXd cam_T_tag(3, 4);  // 3x4 transformation matrix
+        cam_T_tag << m[0][0], m[0][1], m[0][2], cam_T_tag_xpos, m[1][0], m[1][1], m[1][2], cam_T_tag_ypos, m[2][0], m[2][1], m[2][2], cam_T_tag_zpos;
+
+        // 4x1 (X,Y,Z,1) for each tag corner
+        Eigen::MatrixXd cornertl(4, 1);  
```

Comment:
These lines of code are copy pasted with minor changes. Add a TODO to figure out how to unify it / minimise the lines of code